### PR TITLE
feat: support processor virtual filename

### LINKED ISSUE #393: ENOTDIR with virtual filenames from processors
**What version of `eslint` are you using?** 7.21.0

**What version of `prettier` are you using?** 2.2.1

**What version of `eslint-plugin-prettier` are you using?** 3.3.1

```json
{
  "script": {
    "test": "eslint README.md"
  },
  "devDependencies": {
    "eslint": "^7.21.0",
    "eslint-config-prettier": "^8.1.0",
    "eslint-plugin-markdown": "^2.0.0",
    "eslint-plugin-prettier": "^3.3.1",
    "prettier": "^2.2.1"
  }
}
```

**Please paste any applicable config files that you're using (e.g. `.prettierrc` or `.eslintrc` files)**

No `.prettierrc`.

```js
// .eslintrc.js
module.exports = {
    root: true,
    extends: ['plugin:prettier/recommended', 'plugin:markdown/recommended'],
};
```

**What source code are you linting?**

````md
# README.md

```js
var x = 123;
```
````

**What did you expect to happen?**

`npm test` should run without throwing an exception.

**What actually happened?**

```
$ npm test
Oops! Something went wrong! :(

ESLint: 7.21.0

Error: ENOTDIR: not a directory, stat '.../README.md/0_0.js'
Occurred while linting .../README.md/0_0.js:1
    at Object.statSync (node:fs:1128:3)
    at isTypeSync (.../node_modules/prettier/third-party.js:9748:46)
    at getDirectorySync (.../node_modules/prettier/third-party.js:9802:62)
    at ExplorerSync.searchSync (.../node_modules/prettier/third-party.js:9950:66)
    at _resolveConfig (.../node_modules/prettier/index.js:25592:50)
    at Function.resolveConfig.sync (.../node_modules/prettier/index.js:25621:42)
    at Program (.../node_modules/eslint-plugin-prettier/eslint-plugin-prettier.js:167:40)
    at .../node_modules/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (.../node_modules/eslint/lib/linter/safe-emitter.js:45:38)
```

Credit to https://github.com/eslint/eslint-plugin-markdown/issues/174 for originally reporting this. My examples are based on that issue.

`eslint-plugin-markdown` v2 switched to use the [new processor API](https://github.com/eslint/rfcs/tree/master/designs/2018-processors-improvements) that was released as part of ESLint v6. Instead of returning each extracted code block as a string, processors can now return both the source code `text` and a `filename`. In the example code above, the code block's returned `filename` would be `0.js`, with the extension coming from the code block's syntax tag.

The new processor API uses the code block's `filename` to construct a [virtual filename](https://github.com/eslint/rfcs/tree/master/designs/2018-processors-improvements#virtual-filenames-for-code-blocks) as `parentFilename + path.sep + index + filenameFromPreprocess`. In the example above, the code block's virtual filename becomes `README.md/0_0.js`.

Based on the stack trace and a bit of stepping in a debugger, it looks like the exception is thrown when the rule [tries to load the `.prettierrc` for the file](https://github.com/prettier/eslint-plugin-prettier/blob/7d5d3ddbabd3b6e9e2347879889dfef0acaeb7b1/eslint-plugin-prettier.js#L127). If I set `usePrettierrc: false`, a very similar exception is thrown in a call [a few lines later](https://github.com/prettier/eslint-plugin-prettier/blob/7d5d3ddbabd3b6e9e2347879889dfef0acaeb7b1/eslint-plugin-prettier.js#L132).

This exception would occur with any processor that uses the new processor API and yields virtual filenames.

Depending on what it might take to handle code blocks with virtual filenames, https://github.com/eslint/eslint/issues/11989 may be a related issue.

### PR DESCRIPTION
**Pull request by JounQin**
close #393

### CODE REVIEW

**Comment by JounQin** [comment]
@fisker @lydell @BPScott 

Ready for review now!

**Comment by fisker** [comment]
I'm not maintainer of this repo.

I was surprised that this plugin requires file exists to work? Shouldn't `prettier.resolveConfig` work for non-existing files too?

**Comment by JounQin** [comment]
`resolveConfig: true` is parsed into `prettier.resolveConfig`, so `cosmiconfig` will be used.

**Comment by fisker** [comment]
<details>
<summary>Why I can get all info needed?</summary>

```text
> require('prettier').resolveConfig.sync('non-exists-file.foo', {resolveConfig: true})
{
  bracketSpacing: false,
  htmlWhitespaceSensitivity: 'ignore',
  semi: true,
  singleQuote: true,
  trailingComma: 'es5'
}
> require('prettier').resolveConfig.sync('non-exists-file.js', {resolveConfig: true})
{
  bracketSpacing: false,
  htmlWhitespaceSensitivity: 'ignore',
  semi: true,
  singleQuote: true,
  trailingComma: 'es5',
  parser: 'babel'
}
```

</details>

**Comment by fisker** [comment]
Oh, sorry. It's about "virtual filenames from processors"... I don't know anything... I'm not maintainer of this repo. unsubscribing

**Comment by JounQin** [comment]
I don't konw why it throws `ENOTDIR` but not `ENOENT` when using virtual filename...

**Comment by nonameolsson** [comment]
Good job! This resolved my issues, works great so far!

File: eslint-plugin-prettier.js [hunk 1]
```diff
@@ -55,6 +61,26 @@ function reportDifference(context, difference) {
...
+ */
+function normalizeFilepath(filepath) {
+  try {
+    if (fs.statSync(filepath).isFile()) {
+      return filepath;
+    }
+  } catch (err) {
+    // https://github.com/eslint/eslint/issues/11989
+    if (err.code === 'ENOTDIR') {
+      return normalizeFilepath(path.dirname(filepath));
```
Comment (JounQin) [inline eslint-plugin-prettier.js:77]: It could be recursive as described at https://github.com/eslint/eslint/pull/14227#discussion_r602802758

**Comment by BPScott** [comment]
Hi, sorry for the delay in getting to review this PR, my work and peronal lives have been booth pretty busy recently, but I've not forgotten about this. I'll try and get to it soon.

**Review by BPScott — APPROVED** [review]
Thanks for the PR!

Code logic looks good to me, though I'll probably tweak some of the comments

**Comment by JounQin** [comment]
> Thanks for the PR!
> 
> Code logic looks good to me, though I'll probably tweak some of the comments

Thanks, I'm not a native English speaker. 😂

Hope it'll be released ASAP.

**Comment by BPScott** [comment]
@JounQin your english was good! I wanted to add a [some extra explanations](https://github.com/prettier/eslint-plugin-prettier/commit/2ee9363b524857bace01006e3e1d93187cef1516) once I understood how eslint handles virtual filenames.

**Pull request merged**